src/metrics/gc.cc

```diff
@@ -0,0 +1,202 @@
+    kGCCallbackScheduleIdleGarbageCollection = 1 << 6,
+  };
+ */
+
+namespace ao { namespace metrics { namespace gc {
+using namespace v8;
+
+void set_histogram_values(hdr_histogram*, Local<Object>);
+
+typedef struct GCData {
+	uint64_t gcTime;
+  uint64_t gcCount;
+  uint64_t majorCount;
+  uint64_t minorCount;
+} GCData_t;
+
+struct hdr_histogram* h_major;
+struct hdr_histogram* h_minor;
+
+GCData_t raw;
+GCData_t interval_base;
+GCData_t interval;
+
+std::map<std::string, double> PERCENTILES = {
+  {"p50", 50.0}, {"p75", 75.0}, {"p90", 90.0}, {"p95", 95.0}, {"p99", 99.0}
+};
+
+bool enabled = false;
+
+uint64_t gcStartTime;
+
+static NAN_GC_CALLBACK(recordBeforeGC) {
+  gcStartTime = uv_hrtime();
+}
+
+NAN_GC_CALLBACK(afterGC) {
+  // convert elapsed time to μ seconds
+  uint64_t et = (uv_hrtime() - gcStartTime + 500) / 1000;
+
+  // build histogram for elapsed times. scavenges are minor and
+  // mark/sweep/compact are major. the other two types are just
+  // phases of a major garbage collection and are typically very
+  // low values that result in a large standard deviation and lower
+  // mean if added into major garbage collections.
+  // https://stackoverflow.com/questions/60171534/what-do-v8s-incrementalmarking-and-processweakcallbacks-garbage-collections-do/60172044#60172044
+  if (type & kGCTypeMarkSweepCompact) {
+    raw.majorCount += 1;
+    interval.majorCount += 1;
+    hdr_record_values(h_major, et, 1);
+  } else if (type & kGCTypeScavenge) {
+    raw.minorCount += 1;
+    interval.minorCount += 1;
+    hdr_record_values(h_minor, et, 1);
+  } else {
+    // don't update counts or time
+    return;
+  }
+
+  // keep raw up-to-date
+  raw.gcCount += 1;
+  raw.gcTime += et;
+
+  // keep the interval data up-to-date
+  interval.gcCount += 1;
+  interval.gcTime += et;
+}
+
+
+bool initialize() {
+  return true;
+}
+
+//
+// start recording GC events
+//
+bool start() {
+  if (enabled) {
+    return false;
+  }
+  enabled = true;
+
+  // interval base is data at last fetch of interval information so
+  // the first fetch is everything before.
+  memset(&interval_base, 0, sizeof(interval_base));
+  memset(&interval, 0, sizeof(interval));
+
+  hdr_init(1, INT64_C(3600000000), 3, &h_major);
+  hdr_init(1, INT64_C(3600000000), 3, &h_minor);
+
+  Nan::AddGCPrologueCallback(recordBeforeGC);
+  Nan::AddGCEpilogueCallback(afterGC);
+
+  return true;
+}
+
+bool stop () {
+  int status = enabled;
+
+  if (enabled) {
+    enabled = false;
+    memset(&raw, 0, sizeof(raw));
+    memset(&interval_base, 0, sizeof(interval_base));
+    memset(&interval, 0, sizeof(interval));
+    Nan::RemoveGCPrologueCallback(recordBeforeGC);
+    Nan::RemoveGCEpilogueCallback(afterGC);
+    hdr_close(h_major);
+    h_major = nullptr;
+    hdr_close(h_minor);
+    h_minor = nullptr;
+  }
+
+  // if it wasn't enabled return false indicating so, else true.
+  return status;
+}
+
+bool getInterval(const v8::Local<v8::Object> obj) {
```

Comment:
i think they are being passed by reference but i guess i should check the code. i'm not positive what a v8::local looks like now that you mention it; i've presumed they were pointers. i'll take a look at this and let you know what i find.

Reply:
based on the code, and i might not be interpreting it correctly, but it looks like it's being passed by reference.

```
    3aab:	e8 70 f6 ff ff       	callq  3120 <v8::Object::New(v8::Isolate*)@plt>
    3ab0:	48 89 c7             	mov    %rax,%rdi
    3ab3:	49 89 c5             	mov    %rax,%r13
    3ab6:	e8 75 f4 ff ff       	callq  2f30 <ao::metrics::eventloop::getInterval(v8::Local<v8::Object>)@plt>
```
each time it gets an object, moves it to %rdi to pass it to the getInterval(), saves %rax because getInterval() might use it, then calls getInterval(). I'll spend a bit of time in v8 to triangulate this.

Reply:
but because the handle holds the object reference it should always be passed by value, according to this:

https://v8docs.nodesource.com/node-10.15/de/deb/classv8_1_1_local.html

but the code is passing a single pointer. and that seems inconsistent. thoughts?